constraint-evaluation-generator/src/substitution.rs

```diff
@@ -241,93 +219,144 @@ impl Substitutions {
+                    let mut current_base_row = current_base_row_slice.row(0).to_owned();
+                    let next_base_row = next_base_row_slice.row(0);
+                    #(
+                        let det_col = section_row.get_mut(#indices).unwrap();
+                        *det_col = #substitutions;
+                        current_base_row.append(Axis(0), array![*det_col].view()).unwrap();
+                    )*
+                });
         )
     }
 
     fn ext_single_row_substitutions(
-        indices: &[usize],
+        section_start_index: usize,
         substitutions: &[TokenStream],
     ) -> TokenStream {
-        assert_eq!(indices.len(), substitutions.len());
+        let num_substitutions = substitutions.len();
+        let indices = (0..substitutions.len()).collect_vec();
         if indices.is_empty() {
             return quote!();
         }
         quote!(
-            for row_idx in 0..master_base_table.nrows() - 1 {
-                let base_row = master_base_table.row(row_idx);
-                let mut extension_row = master_ext_table.row_mut(row_idx);
-                #(
-                let (ext_row, mut det_col) =
-                    extension_row.multi_slice_mut((s![..#indices],s![#indices..=#indices]));
-                det_col[0] = #substitutions;
-                )*
-            }
+            let (original_part, mut current_section) = master_ext_table.multi_slice_mut(
+                (
+                    s![.., 0..#section_start_index],
+                    s![.., #section_start_index..#section_start_index+#num_substitutions],
+                )
+            );
+            Zip::from(master_base_table.rows())
+                .and(original_part.rows())
+                .and(current_section.rows_mut())
+                .par_for_each(
+                    |base_table_row, original_row, mut section_row| {
+                        let mut extension_row = original_row.to_owned();
+                        #(
+                            let (original_row_extension_row, mut det_col) =
+                                section_row.multi_slice_mut((s![..#indices],s![#indices..=#indices]));
+                            det_col[0] = #substitutions;
+                            extension_row.append(Axis(0), array![det_col[0]].view()).unwrap();
+                        )*
+                    }
+                );
         )
     }
 
-    fn ext_dual_row_substitutions(indices: &[usize], substitutions: &[TokenStream]) -> TokenStream {
-        assert_eq!(indices.len(), substitutions.len());
+    fn ext_dual_row_substitutions(
+        section_start_index: usize,
+        substitutions: &[TokenStream],
+    ) -> TokenStream {
+        let num_substitutions = substitutions.len();
+        let indices = (0..substitutions.len()).collect_vec();
         if indices.is_empty() {
             return quote!();
         }
         quote!(
-            for curr_row_idx in 0..master_base_table.nrows() - 1 {
```

Comment:
Ah, the power of copy & paste, where base tables appear in extension table code.

Is there some way to merge the functionality? It seems redundant to mention things twice. If there's a good reason, it might be worth a doc comment for this function. I'm aware the maybe-code-duplication is not your doing, but this seems like a good time to look at it.

Reply:
Perfect, thanks for the insight. :ok_hand: